### ISSUE TITLE
Issues integrating Keras with spearmint

### ISSUE THREAD

**Issue by akmahaja** [fchollet/keras#2266]
When i  am using spearmint to optimize the hyperparamaters for Keras models, for the first time it runs fine. But the second job onwards it always throws the following error. 

<type 'exceptions.TypeError'>, TypeError('An update must have the same type as the original shared variable (shared_var=<TensorType(float32, matrix)>, shared_var.type=TensorType(float32, matrix), update_val=Elemwise{add,no_inplace}.0, update_val.type=TensorType(float64, matrix)).', 'If the difference is related to the broadcast pattern, you can call the tensor.unbroadcast(var, axis_to_unbroadcast[, ...]) function to remove broadcastable dimensions.'), <traceback object at 0x18a5c5710>)

I am using the following code to load the pre-created numpy matrices of the train data and test data. The following params are passed by the optimization python script. But the set of parameters work fine if run without spearmint.  

    def load_train_data(arg_type, params=None):

        X_train1 = pickle.load(open(arg_type+"_train1","rb"))

         X_train2 = pickle.load(open(arg_type+"_train2","rb"))

        Y_train = pickle.load(open(arg_type+"_train_labels","rb"))

         model=combined_model(X_train1,X_train2,Y_train,params)

        X_test1 = pickle.load(open(arg_type+"_test1","rb"))

        X_test2 = pickle.load(open(arg_type+"_test2","rb"))

         Y_test = pickle.load(open(arg_type+"_test_labels","rb"))

        loss = model.evaluate({'input1': X_test1,'input2': X_test2,'output':Y_test},batch_size=450)

        return loss

**Comment by kevinbache** [comment]
Spearmint passes parameter values as singleton arrays, so instead of e.g.: batch_size=2 you get batch_size=np.array([2]).  See if converting your spearmint-originated values to basic python types with float(), int(), and bool() helps.

**Comment by akmahaja** [comment]
Thank you this seemed to have solved the issue. Even though I was extracting the values for the 0 position in a list, somehow explicitly converting them to float, helped.

**Comment by nouiz** [comment]
A quick explanation, Theano will use the smallest possible dtype for python
int and float. But for numpy, we respect the givent dtype.

So it was a numpy array for float64 dtype, and so it was causing upcast to
float64 in the graph. But when you convert it to a python float, Theano
will use the smallest possible dtype without loss of precission, so we will
downcast it to float32. So it won't cause upcast to float64 in the graph.

**Comment by alyato** [comment]
hi @akmahaja  Do you share one example how to use the spearmint in image classify. Thanks

**Comment by akmahaja** [comment]
@alyato Spearmint is used for hyper parameter optimization. Not sure what you mean by using the spearmint in image classify. Please clarify.

**Comment by alyato** [comment]
@akmahaja yeah. I want to use spearmint in face recognation,but i don't know how to set the config of spearmint.Do you give me one example ? Thanks
